contracts/RestrictedMultisigBase.sol

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.23;
+
+abstract contract RestrictedMultisigBase {
+    error IsNotMember();
+    error SenderIsNotMember();
+    error SenderIsNotOwner();
+    error DataIsNotEqual();
+
+    struct Action {
+        uint256 actionType;
+        bytes data;
+        bool isExecuted;
+        address[] signers;
+    }
+
+    address public owner;
+
+    address[] public membersList;
+    mapping(address => bool) public members;
+    uint256 public quorum;
+
+    mapping(bytes32 actionHash => Action) actions;
+    mapping(address signer => mapping(bytes32 actionHash => bool support)) public approves;
+
+    constructor(address _owner, address[] memory _members, uint256 _quorum) {
+        quorum = _quorum;
+        owner = _owner;
+
+        for (uint256 i = 0; i < _members.length; ++i) {
+            _addMember(_members[i]);
+        }
+    }
+
+    function _vote(Action memory _action, bool _supports) internal {
+        bytes32 actionHash = _hashAction(_action);
+        if (actions[actionHash].data.length == 0) {
+            actions[actionHash].actionType = _action.actionType;
+            actions[actionHash].data = _action.data;
+        } else {
+            _checkStoredAction(_action);
```

Comment:
why is this needed given that we already hash the action when storing, and the action type+data is immutable?